internal/operator/uninstall.go

```diff
@@ -69,82 +81,116 @@ func (u *Uninstall) Run(ctx context.Context) error {
-	if err != nil {
-		return err
+	var crds, csvs, others []controllerutil.Object
+	if sub.Status.InstallPlanRef != nil {
+		ipKey := types.NamespacedName{
+			Namespace: sub.Status.InstallPlanRef.Namespace,
+			Name:      sub.Status.InstallPlanRef.Name,
+		}
+		var err error
+		crds, csvs, others, err = u.getInstallPlanResources(ctx, ipKey)
+		if err != nil {
+			return fmt.Errorf("get install plan resources: %v", err)
+		}
 	}
 
 	// Delete the subscription first, so that no further installs or upgrades
 	// of the operator occur while we're cleaning up.
-	if err := u.config.Client.Delete(ctx, sub); err != nil {
-		return fmt.Errorf("delete subscription %q: %v", sub.Name, err)
-	}
-	u.config.Log("subscription %q deleted\n", sub.Name)
-
-	// Ensure CustomResourceDefinitions are deleted first, so that the operator
-	// has a chance to handle CRs that have finalizers.
-	sort.SliceStable(deleteObjs, func(i, j int) bool {
-		return deleteObjs[i].GetObjectKind().GroupVersionKind().Kind == "CustomResourceDefinition"
-	})
-	for _, obj := range deleteObjs {
-		err := u.config.Client.Delete(ctx, obj)
-		if err != nil && !apierrors.IsNotFound(err) {
+	if err := u.deleteObjects(ctx, false, sub); err != nil {
+		return err
+	}
+
+	if u.DeleteCRDs {
+
+		// Ensure CustomResourceDefinitions are deleted first, so that the operator
+		// has a chance to handle CRs that have finalizers.
+		if err := u.deleteObjects(ctx, true, crds...); err != nil {
 			return err
 		}
-		if err == nil {
-			u.config.Log("%s %q deleted\n", strings.ToLower(obj.GetObjectKind().GroupVersionKind().Kind), obj.GetName())
-		}
+	}
+
+	// Delete CSVs and all other objects created by the install plan.
+	objects := append(csvs, others...)
+	if err := u.deleteObjects(ctx, false, objects...); err != nil {
+		return err
 	}
 
 	// Delete the catalog source. This assumes that all underlying resources related
 	// to this catalog source have an owner reference to this catalog source so that
 	// they are automatically garbage-collected.
-	if err := u.config.Client.Delete(ctx, catsrc); err != nil {
-		return fmt.Errorf("delete catalog source: %v", err)
+	if err := u.deleteObjects(ctx, false, catsrc); err != nil {
+		return err
 	}
-	u.config.Log("catalogsource %q deleted\n", catsrc.Name)
 
 	// If this was the last subscription in the namespace and the operator group is
 	// the one we created, delete it
-	if len(subs.Items) == 1 {
+	if u.DeleteOperatorGroups && len(subs.Items) == 1 {
 		ogs := v1.OperatorGroupList{}
 		if err := u.config.Client.List(ctx, &ogs, client.InNamespace(u.config.Namespace)); err != nil {
 			return fmt.Errorf("list operatorgroups: %v", err)
 		}
 		for _, og := range ogs.Items {
-			if og.GetName() == SDKOperatorGroupName {
-				if err := u.config.Client.Delete(ctx, &og); err != nil {
-					return fmt.Errorf("delete operatorgroup %q: %v", og.Name, err)
+			if len(u.DeleteOperatorGroupNames) == 0 || slice.ContainsString(u.DeleteOperatorGroupNames, og.GetName(), nil) {
```

Comment:
```suggestion
                         og := og
			if len(u.DeleteOperatorGroupNames) == 0 || slice.ContainsString(u.DeleteOperatorGroupNames, og.GetName(), nil) {
```

Reply:
also... expect a PR soonish with linter fixes

Reply:
Maybe I misundestand the problem, but re-scoping the `og` variable shouldn't be necessary when
- the value of the variable is only used inside the scope of the loop and never assigned to something that has scope outside the loop AND
- the value is not used in a goroutine declared inside the loop (unless passing it as a re-scoped argument to the goroutine function)

I'll go ahead and make this change because it doesn't hurt anything and may prevent this issue popping up later if the loop logic changes.